Group capture

ratierd opened the pull request:
Hi ! Thanks for creating this tool. Me and my team would like to use this project to enforce a rule so that every file in a given folder would be named after the folder name.

In order to have something like this : 

- folder
  - index.ts // Allowed
  - folder.view.ts // Allowed
  - folder.hooks.ts // Allowed
  - folder.specs.ts // Allowed
  - foo.ts // Forbidden

I've created this PR in order to achieve this but I think the feature could be used in other scenarios.

It relies on micromatch capture feature which allows to identify group in the glob. I then use it in the rule using a special  syntax `$index` referring to a given capture group.

DukeLuo commented:
@ratierd  Thanks for your pull request. Group capture is a great idea, it's very useful for rule `filename-naming-convention`.

DukeLuo commented on tests/lib/rules/filename-naming-convention.posix.js:
```diff
@@ -1271,3 +1271,95 @@ ruleTester.run(
 );
+
+ruleTester.run(
+  "filename-naming-convention with option: [{ '**/components/featureA/*.*': '$1*' }]",
```
Each ruleTester should always have the same options in this project. I noticed there are different options in this ruleTest.

Please let me know if you have any questions or concerns.

DukeLuo commented on tests/lib/rules/filename-naming-convention.posix.js:
```diff
@@ -1271,3 +1271,95 @@ ruleTester.run(
+          { ignoreMiddleExtensions: true },
+        ],
+      },
+    ],
```
As shown in this test case and the one above, `ignoreMiddleExtensions` seems cant work well now.

DukeLuo reviewed (changes requested):
I really like the idea of group capture. Thank you for your contribution to this project. I have gone through your pull request and I do have a few concerns that I would like to address:
1. Regarding the `ignoreIndexFiles` parameter, I think we can exclude the index file using the glob expression instead of introducing a new parameter.

2. I noticed the use of `$1*` in the code, and I was wondering about its meaning and how it differs from `$1`.  Is `$1` enough?

3. We use `$` to indicate our custom syntax, but what if we need to include a `$` in a glob expression? How do we handle this case?

Please let me know if you have any questions or if there is anything I can do to assist you in addressing these concerns.

ratierd commented:
> I really like the idea of group capture. Thank you for your contribution to this project. I have gone through your pull request and I do have a few concerns that I would like to address:
> 
> 1. Regarding the `ignoreIndexFiles` parameter, I think we can exclude the index file using the glob expression instead of introducing a new parameter.
> 2. I noticed the use of `$1*` in the code, and I was wondering about its meaning and how it differs from `$1`.  Is `$1` enough?
> 3. We use `$` to indicate our custom syntax, but what if we need to include a `$` in a glob expression? How do we handle this case?
> 
> Please let me know if you have any questions or if there is anything I can do to assist you in addressing these concerns.

1. Yes I though so, I replaced it with something more generic, ignorePattern. Leveraging the ignore API of micromatch
2. I tried to avoid having the extra * but it throws an error `There is an invalid pattern "featureA", please check it` since it is not a glob or a builtin pattern
3. I didn't think about that. I updated the syntax to be `<index>` instead. Since angle brackets are illegal in Windows filename and isn't a glob special character it should be safe

DukeLuo commented on docs/rules/filename-naming-convention.md:
````diff
@@ -98,6 +121,28 @@ module.exports = {
+  },
+};
+```
+
````
There is no necessary to introduce a new parameter `ignorePattern`, you can use a glob expression like `src/*/!(index).*` to select target files. [see more](https://globster.xyz/?q=src%2F*%2F!(index).*&f=src%2FPortal%2FPortal.test.tsx%2Csrc%2FPortal%2FPortal.tsx%2Csrc%2FPortal%2FP.py%2Csrc%2FPortal%2FPortal.types.tsx%2Csrc%2FPortal%2Findex.tsx%2Csrc%2FPortal%2Finde.css)

ratierd replied:
Okey, yes it's even better . Thanks

DukeLuo commented on lib/rules/filename-naming-convention.js:
```diff
@@ -45,8 +46,24 @@ module.exports = {
+          context.options[0],
+          filenameWithPath
+        );
+
```
You get a error an error `There is an invalid pattern "featureA", please check it since it is not a glob or a builtin pattern` since you transform custom syntax firstly and then run into the `checkSettings` validation function. In this case, `<1>` will be `featureA`, but `featureA` is not a valid pattern.

I have an idea, we only allow three kinds of pattern styles for this rule: builtin pattern, glob pattern and our group capture syntax(`<index>` or `$index`). We only allow user choose one style at one time, and dont allow mix with them. So in this way, `<1>*`, `CAMEL_CASE*` both are illegal.

I feel like this would reduce the complexity, and make it more clear and easy to use.

ratierd replied:
Yes it seems like a good idea.

I think I got most of it down but  I can't figure out what part of the coverage I'm missing.
I use sonar in some other projects to see which lines specifically aren't covered.
Do you use something similar ?

DukeLuo replied:
> Yes it seems like a good idea.
> 
> I think I got most of it down but I can't figure out what part of the coverage I'm missing. I use sonar in some other projects to see which lines specifically aren't covered. Do you use something similar ?

Check out [this link](https://app.codecov.io/gh/DukeLuo/eslint-plugin-check-file/tree/group_capture).

DukeLuo reviewed (approved):
Cool, looks good to me. Thanks for your PR. It will be released in next version.

DukeLuo commented:
This PR has been released.
